### PR TITLE
Enable Cray compiler on Theta

### PR DESCRIPTION
**Pull request by amametjanov**
Enable Cray compiler on Theta

[BFB]

### CODE REVIEW

**Comment by amametjanov** [comment]
This needs the following PRs to avoid various build issues:
1. https://github.com/E3SM-Project/scorpio/pull/328
2. https://github.com/NGEET/fates/pull/671
3. https://github.com/MPAS-Dev/MPAS-Model/pull/629

**Comment by sarats** [comment]
@amametjanov I presume you turned off optimization for certain files due to compilation times? 

A suggestion: Move the noopt files to a platform specific Depends file instead of compiler. I tested the existing two files(cice/src/mpi/ice_boundary.F90,cice/src/serial/ice_boundary.F90) on Tulip and they didn't have expensive compilation times.

File: cime/config/e3sm/machines/config_compilers.xml [hunk 1]
```diff
@@ -86,29 +86,32 @@ flags should be captured within MPAS CMake files.
...
   <FC_AUTO_R8>
     <base> -s real64 </base>
   </FC_AUTO_R8>
   <FFLAGS>
-    <base> -O2  -f free -N 255  -h byteswapio -em </base>
-    <append compile_threaded="FALSE"> -h noomp </append>
-    <append DEBUG="TRUE"> -g -trapuv  -Wuninitialized </append>
+    <base> -f free -N 255 -h byteswapio -em </base>
+    <append compile_threaded="TRUE"> -h omp </append>
+    <append compile_threaded="FALSE"> -M1077 </append>
```
Comment (sarats) [inline cime/config/e3sm/machines/config_compilers.xml:102]: What's this flag for?

File: cime/config/e3sm/machines/config_compilers.xml [hunk 2]
```diff
@@ -86,29 +86,32 @@ flags should be captured within MPAS CMake files.
...
+    <append DEBUG="TRUE"> -g -O0 </append>
   </FFLAGS>
   <FFLAGS_NOOPT>
     <base> -O0 </base>
   </FFLAGS_NOOPT>
   <HAS_F2008_CONTIGUOUS>TRUE</HAS_F2008_CONTIGUOUS>
   <LDFLAGS>
     <base> -Wl,--allow-multiple-definition -h byteswapio </base>
-    <append compile_threaded="FALSE"> -h noomp </append>
+    <append compile_threaded="TRUE"> -h omp </append>
```
Comment (sarats) [inline cime/config/e3sm/machines/config_compilers.xml:111]: Just trying to understand: What's the motivation for the switch for noomp to omp? The old version didn't have any threaded builds?

File: cime/config/e3sm/machines/config_compilers.xml (hunk 1 above)
Comment (amametjanov) [reply cime/config/e3sm/machines/config_compilers.xml:102]: It's to remove compiler warnings about presence of `!$omp` pragmas in pureMPI compilation mode.

File: cime/config/e3sm/machines/config_compilers.xml (hunk 2 above)
Comment (amametjanov) [reply cime/config/e3sm/machines/config_compilers.xml:111]: `noomp` is compiler's default, turning on `omp` only for threaded builds.

File: cime/config/e3sm/machines/Depends.cray.cmake [hunk 3]
```diff
@@ -1 +1 @@
-list(NOOPT_FILES APPEND cice/src/mpi/ice_boundary.F90 cice/src/serial/ice_boundary.F90)
+list(APPEND NOOPT_FILES cice/src/mpi/ice_boundary.F90 cice/src/serial/ice_boundary.F90 cice/src/source/ice_domain.F90 cam/src/physics/cam/gw_common.F90)
```
Comment (amametjanov) [inline cime/config/e3sm/machines/Depends.cray.cmake:1]: Having `ice_domain.F90` and `gw_common.F90` at `-O2` or `-O1` was producing non-BFB results in ERS tests of the `e3sm_integration` test suite. Compiling them with `-O0` made the tests BFB.

**Comment by sarich** [comment]
@amametjanov  @sarats 

is this ready to start merging?

**Comment by amametjanov** [comment]
Let me rebase and retest after Theta comes back from PM.

**Comment by rljacob** [comment]
@amametjanov what is the status of this PR?

**Comment by amametjanov** [comment]
Waiting for cray modules to be re-installed. Closing until then.

**Pull request closed without merging**